### DISCUSSION TITLE
Can I send a SUBSCRIBE command using a context that's blocked on redisGetReply

### DISCUSSION THREAD

**Discussion by peter-lee-schneider** [redis/hiredis#1078] [Q&A] [answered]
I see that "a redisContext is not thread-safe" according to <hiredis/README.md>, but I also see the following comment in <https://redis.io/commands/subscribe/>:
- Once the client enters the subscribed state it is not supposed to issue any other commands, except for additional SUBSCRIBE

I tried running the following test and it seems to work for me:
- I have one thread that opens a hiredis context using the Synchronous API, then calls redisCommand(SUBSCRIBE) on a channel, and then calls redisGetReply() to wait for a notification.
- I have a second thread that subscribes to additional channels using the same hiredis context from the first thread, where I notice that redisCommand(SUBSCRIBE) returns success

Now, does hiredis support a call to redisCommand(SUBSCRIBE) on a context that's blocked on redisGetReply()?
If not, any suggestions on how I can subscribe to additional channels on a context that is already waiting for a notification?

Alternatively, I have looked into calling redisSetTimeout() before redisGetReply(), however after calling redisReconnect() I noticed that all the subscribed channels were reset/gone. It would be messy for me to try and re-subscribe to all the channels again.

I have not looked into the Asynchronous API yet, as the Synchronous API fits in much better with our application.

Thank you.

**Comment by michael-grunder** [comment, marked as answer]
Hi, thanks for the question.

Although this may work sometimes, it will inevitably lead to undefined behavior and fail in strange and difficult-to-predict ways.

The reason for this is that two independent threads will be attempting to operate on the same shared internal context and hiredis does not do any synchronization internally.

The async API can do what you need, and we actually recently merged a polling adapter, which would allow you to use it without having to add an event library to your application.

[The adapter in question](https://github.com/redis/hiredis/blob/master/adapters/poll.h)
[An example program that uses it](https://github.com/redis/hiredis/blob/master/examples/example-poll.c)

Another solution could possibly be to `PUBLISH` a message to the blocked client with some kind of special payload that would inform that client to subscribe to an additional/different channel.

**Comment by peter-lee-schneider** [comment]
Hi Michael,

Thank you for your response, that really clears things up!

I like your suggestion on using another PUBLISH message, which I'll look into further.
